src/Message/AIMRefundRequest.php

```diff
@@ -54,15 +60,71 @@ public function send()
         /** @var AIMResponse $response */
         $response = parent::send();
 
-        if (!$response->isSuccessful() && $this->shouldVoidIfRefundFails() &&
-            $response->getReasonCode() == AIMResponse::ERROR_RESPONSE_CODE_CANNOT_ISSUE_CREDIT
-        ) {
-            // This transaction has not yet been settled, hence cannot be refunded. But a void is possible.
-            $voidRequest = new CIMVoidRequest($this->httpClient, $this->httpRequest);
-            $voidRequest->initialize($this->getParameters());
-            $response = $voidRequest->send();
+        if (!$response->isSuccessful()) {
+            $shouldVoid = $this->shouldVoidIfRefundFails();
+            $isCreditIssueError = $response->getReasonCode() == AIMResponse::ERROR_RESPONSE_CODE_CANNOT_ISSUE_CREDIT;
+            if ($shouldVoid && $isCreditIssueError) {
+                // This transaction has not yet been settled, hence cannot be refunded. But a void is possible.
+                $voidRequest = new CIMVoidRequest($this->httpClient, $this->httpRequest);
+                $voidRequest->initialize($this->getParameters());
+                $response = $voidRequest->send();
+            } else {
+                $errorCode = null;
+                if ($response->getData() && isset($response->getData()->messages->message)) {
+                    $errorCode = (string) $response->getData()->messages->message->code;
+                }
+
+                if ($errorCode === self::ERROR_CODE_INVALID_PAYMENT) {
+                    /** @var QueryDetailResponse $transactionResponse */
+                    $transactionResponse = $this->getTransactionDetails();
+
+                    if (!$transactionResponse->isSuccessful()) {
+                        return $response;
+                    }
+
+                    $transactionData = $transactionResponse->getTransaction();
+                    $hasCustomerProfileId = isset($transactionData['profile']['customerProfileId']);
```

Comment:
You are inside an AIM class -- AIM does not have the concept of customer profiles -- only CIM has that, right? Need to keep those classes separate.

Reply:
yes that's right, CIMRefundRequest only had the declaration extending AIM class of class. So added it here. We can move it to CIMRefundRequest.php
`class CIMRefundRequest extends AIMRefundRequest
{

}
`